resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs

```diff
@@ -0,0 +1,69 @@
+using JetBrains.Application.Threading;
+using JetBrains.Application.Threading.Tasks;
+using JetBrains.Collections.Viewable;
+using JetBrains.Lifetimes;
+using JetBrains.ProjectModel;
+using JetBrains.ProjectModel.Caches;
+using JetBrains.ProjectModel.Tasks;
+using JetBrains.ReSharper.Feature.Services.Cpp.Caches;
+using JetBrains.ReSharper.Plugins.Unity.ProjectModel;
+using JetBrains.Util;
+using JetBrains.Util.Threading.Tasks;
+
+namespace JetBrains.ReSharper.Plugins.Unity.HlslSupport
+{
+    [SolutionComponent]
+    public class CgIncludeDirectoryTracker
+    {
+        private const string CG_FOLDER_PATH = "CgIncludeFolderPath";
+
+        public CgIncludeDirectoryTracker(Lifetime lifetime, UnityReferencesTracker unityReferencesTracker,
+            SolutionCaches solutionCaches, IShellLocks shellLocks, ISolutionLoadTasksScheduler scheduler,
+            CppGlobalCacheImpl cppGlobalCache, UnityVersion unityVersion, ILogger logger)
+        {
+            scheduler.EnqueueTask(new SolutionLoadTask("InitCgIncludeDirectoryTracker", SolutionLoadTaskKinds.PreparePsiModules,
+                () =>
+                {
+                    unityReferencesTracker.HasUnityReference.AdviseOnce(lifetime, _ =>
+                    {
+                        if (solutionCaches.PersistentProperties.TryGetValue(CG_FOLDER_PATH, out var result))
+                        {
+                            var oldPath = FileSystemPath.TryParse(result, FileSystemPathInternStrategy.INTERN);
+                            var newPath = GetCgIncludeFolderPath(unityVersion);
+                            if (!oldPath.Equals(newPath))
+                            {
+                                cppGlobalCache.IsInitialUpdateFinished.Change.Advise(lifetime, v =>
+                                {
+                                    if (v.HasNew && v.New)
+                                    {
+                                        shellLocks.Tasks.StartNew(lifetime, Scheduling.MainGuard, TaskPriority.High, () =>
+                                        {
+                                            logger.Verbose("Dropping C++ cache, because Unity version is changed");
+                                            cppGlobalCache.ResetCache();
+                                            solutionCaches.PersistentProperties[CG_FOLDER_PATH] = newPath.FullPath;
```

Comment:
This persistent property isn't used anywhere else. Is it just a marker to know to rebuild the cache?

Reply:
Yes